### PR TITLE
[Core] Register renderers conditionally based on OS version

### PR DESCRIPTION
**Pull request by adrianknight89**
### Description of Change ###

On iOS, `ExportRendererAttribute` allows us to supply two different renderers for MDP on phones and tablets with `UIUserInterfaceIdiom` attribute parameter. We can build upon this to extend the behavior of the exporter such that different renderers can be utilized based on OS version.

We can now do the following:

```
[assembly:ExportRenderer(typeof(Entry), typeof(MyRenderer1), MaximumSdkVersion = 9)]
namespace fgfg789.iOS
{
	public class MyRenderer1 : EntryRenderer
	{
		public MyRenderer1()
		{
			; // Put a breakpoint here. It won't be hit because the renderer 
                          // has not been registered on iOS 10.
		}
	}
}

[assembly: ExportRenderer(typeof(Entry), typeof(MyRenderer2), MinimumSdkVersion = 10)]
namespace fgfg789.iOS
{
	public class MyRenderer2 : EntryRenderer
	{
		public MyRenderer2()
		{
			; // on iOS 10+, This renderer will be registered.
		}
	}
}
```

We can also use more than two renderers if we like so long as we make sure only one renderer is registered per version:

```
[assembly: ExportRenderer(typeof(Entry), typeof(MyRenderer1), MinimumSdkVersion = 6, MaximumSdkVersion = 7)]
[assembly: ExportRenderer(typeof(Entry), typeof(MyRenderer2), MinimumSdkVersion = 8, MaximumSdkVersion = 9]
[assembly: ExportRenderer(typeof(Entry), typeof(MyRenderer3), MinimumSdkVersion = 10)]
```

For further discussion, see [here](https://forums.xamarin.com/discussion/80341/methods-for-registering-renderers-during-runtime-instead-of-using-exportrenderer-attribute#latest).

Note that I do not know how to get OS version in RT (not sure if it's possible either) so I left it out.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/HandlerAttribute.cs [hunk 1]
```diff
@@ -5,19 +5,60 @@ namespace Xamarin.Forms
...
+
+			var isSupported = (bool)FilteredSdkVersionInformation[0];
+			var minimumSdkVersion = (int)FilteredSdkVersionInformation[1];
+			var maximumSdkVersion = (int)FilteredSdkVersionInformation[2];
+
+			if(minimumSdkVersion <= MajorVersion.Value && MajorVersion.Value <= maximumSdkVersion)
+				return isSupported;
+
+			// This means either the OS version has not originally fallen into the supplied range or 
+			// the OS version has changed and the supplied range is no longer valid.
```
Comment (adrianknight89) [inline Xamarin.Forms.Core/HandlerAttribute.cs]: Not sure if we should throw an exception here or return true so that the renderer is registered.

File: Xamarin.Forms.Core/HandlerAttribute.cs [hunk 2]
```diff
@@ -5,19 +5,60 @@ namespace Xamarin.Forms
 	[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
 	public abstract class HandlerAttribute : Attribute
 	{
+		protected HandlerAttribute(Type handler, Type target, object[] filteredSdkVersionInformation) : this(handler, target)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/HandlerAttribute.cs]: we don't do 'magic' `object[]` parameter in public API. it is likely to throw ~~on line 33~~ on `Validate()`, if the array doesn't have the right size, or the worn element types

Also, there's no need to add ctor parameters to the attribute. you can add properties, and those will be available as optional named param when you declare the attribute. That's the beauty, and weirdness, of attributes: attributes had optional parameters long before C# got it, and `DebuggerDisplayAttribute` had string interpolation long before c#6...

Comment (adrianknight89) [inline Xamarin.Forms.Core/HandlerAttribute.cs]: I added a ctor parameter because `UIUserInterfaceIdiom` was added as such. Instead of `object[]`, what would be a proper data type? The reason why I added it in the first place was breaking up `{isSupported, minimumSdkVersion, maximumSdkVersion}` didn't make sense because `isSupported` didn't make sense on its own. Would it be okay if there were three properties (perhaps `isSupported can be named `isFilteredSdkVersionSupported`)?

**Comment by adrianknight89** [comment]
@StephaneDelcroix Can you review again? I'll also work on tests.

**Comment by adrianknight89** [comment]
@StephaneDelcroix Added tests as well.

File: Xamarin.Forms.Platform.WinRT/WindowsDeviceInfo.cs [hunk 3]
```diff
@@ -110,5 +111,20 @@ void OnOrientationChanged(DisplayInformation sender, object args)
 		{
 			CurrentOrientation = GetDeviceOrientation(sender.CurrentOrientation);
 		}
+
+		internal static string GetOSVersion()
+		{
+#if WINDOWS_UWP
```
Comment (samhouts) [inline Xamarin.Forms.Platform.WinRT/WindowsDeviceInfo.cs]: We can get rid of this now.

File: Xamarin.Forms.Platform.WinRT/ExportRendererAttribute.cs [hunk 4]
```diff
@@ -9,10 +9,13 @@ namespace Xamarin.Forms.Platform.WinRT
 #endif
 {
 	[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
-	public sealed class ExportRendererAttribute : HandlerAttribute
+	public sealed class ExportRendererAttribute : BaseExportRendererAttribute
 	{
 		public ExportRendererAttribute(Type handler, Type target) : base(handler, target)
 		{
+#if WINDOWS_UWP
```
Comment (samhouts) [inline Xamarin.Forms.Platform.WinRT/ExportRendererAttribute.cs]: We can get rid of this now.

File: Xamarin.Forms.Platform.iOS/ExportRendererAttribute.cs [hunk 5]
```diff
@@ -23,7 +23,10 @@ public ExportRendererAttribute(Type handler, Type target) : base(handler, target
 
 		public override bool ShouldRegister()
 		{
-			return !Idiomatic || Idiom == UIDevice.CurrentDevice.UserInterfaceIdiom;
+			if (!(!Idiomatic || Idiom == UIDevice.CurrentDevice.UserInterfaceIdiom))
```
Comment (samhouts) [inline Xamarin.Forms.Platform.iOS/ExportRendererAttribute.cs]: A little hard to parse this. Can we make it easier to understand without the double negative? :)

**Comment by adrianknight89** [comment]
The same functionality needs to be implemented in the new projects (Tizen, WPF, GTK, etc.). Let me know if this feature can be merged so I can re-work it.

**Comment by rmarinho** [comment]
@adrianknight89 yes we will be merging this, so if you can get this updated it would be great.

**Comment by rmarinho** [comment]
any luck running the new stuff @adrianknight89  ?

**Comment by adrianknight89** [comment]
Unfortunately no. At this point, I wouldn't mind a team member owning this. Most of the work is done already.

**Comment by rookiejava** [comment]
In tizen, you can get th e platform version as follow : 
`Tizen.System.Information.TryGetValue<string>("tizen.org/feature/platform.version", out platformVersion);
`

For more details, please refer to [API Reference](https://developer.tizen.org/dev-guide/csapi/api/Tizen.System.Information.html#Tizen_System_Information_TryGetValue__1_System_String___0__) or [Developer Guide](https://developer.tizen.org/development/guides/.net-application/device-settings-and-systems/system-and-runtime-information).
